pkg/vm/engine/tae/catalog/base.go

```diff
@@ -22,466 +22,569 @@ import (
+// for replay
+func (be *BaseEntry) GetExactUpdateNode(startts types.TS) (node *UpdateNode) {
+	be.MVCC.Loop(func(n *common.DLNode) bool {
+		un := n.GetPayload().(*UpdateNode)
+		if un.Start == startts {
+			node = un
+			return false
 		}
-		return be.CreateAt
+		// return un.Start < startts
+		return true
+	}, false)
+	return
+}
+
+func (be *BaseEntry) NeedWaitCommitting(startTS types.TS) (bool, txnif.TxnReader) {
+	un := be.GetUpdateNodeLocked()
+	if un == nil {
+		return false, nil
+	}
+	if !un.IsCommitting() {
+		return false, nil
 	}
-	if be.CreateAt != be.Txn.GetCommitTS() {
-		return be.CreateAt
+	if un.Txn.GetStartTS().Equal(startTS) {
+		return false, nil
 	}
-	var cts types.TS
-	return cts
+	return true, un.Txn
 }
 
-func (be *BaseEntry) CloneCreate() *BaseEntry {
-	info := be.PrevCommit.Clone()
-	cloned := &BaseEntry{
-		CommitInfo: *info,
-		ID:         be.ID,
-		CreateAt:   be.CreateAt,
+// active -> w-w/same
+// rollbacked/ing -> next
+// committing shouldRead
+
+// get first
+// get committed
+// get same node
+// get to read
+func (be *BaseEntry) InTxnOrRollbacked() bool {
+	un := be.GetUpdateNodeLocked()
+	if un == nil {
+		return true
 	}
-	return cloned
+	return un.IsActive()
 }
 
-func (be *BaseEntry) Clone() *BaseEntry {
-	info := be.CommitInfo.Clone()
-	cloned := &BaseEntry{
-		CommitInfo: *info,
-		ID:         be.ID,
-		CreateAt:   be.CreateAt,
-		DeleteAt:   be.DeleteAt,
+func (be *BaseEntry) IsDroppedCommitted() bool {
+	un := be.GetUpdateNodeLocked()
+	if un == nil {
+		return false
 	}
-	return cloned
+	if un.IsActive() {
+		return false
+	}
+	return un.HasDropped()
+}
+
+func (be *BaseEntry) PrepareWrite(txn txnif.TxnReader, rwlocker *sync.RWMutex) (err error) {
+	node := be.GetUpdateNodeLocked()
+	if node.IsActive() {
+		if node.IsSameTxn(txn.GetStartTS()) {
+			return
+		}
+		return txnif.ErrTxnWWConflict
+	}
+
+	if node.Start.Greater(txn.GetStartTS()) {
+		return txnif.ErrTxnWWConflict
+	}
+	return
 }
 
-func (be *BaseEntry) WriteTo(w io.Writer) (n int64, err error) {
+func (be *BaseEntry) WriteOneNodeTo(w io.Writer) (n int64, err error) {
 	if err = binary.Write(w, binary.BigEndian, be.ID); err != nil {
 		return
 	}
-	if err = binary.Write(w, binary.BigEndian, be.CreateAt); err != nil {
+	n += 8
+	var n2 int64
+	n2, err = be.GetUpdateNodeLocked().WriteTo(w)
+	if err != nil {
 		return
 	}
-	if err = binary.Write(w, binary.BigEndian, be.DeleteAt); err != nil {
+	n += n2
+	return
+}
+
+func (be *BaseEntry) WriteAllTo(w io.Writer) (n int64, err error) {
+	if err = binary.Write(w, binary.BigEndian, be.ID); err != nil {
 		return
 	}
-	sn := int64(0)
-	if sn, err = be.CommitInfo.WriteTo(w); err != nil {
+	n += 8
+	if err = binary.Write(w, binary.BigEndian, be.length); err != nil {
 		return
 	}
-	n = sn + 8 + 8 + 8
+	n += 8
+	be.MVCC.Loop(func(node *common.DLNode) bool {
+		var n2 int64
+		n2, err = node.GetPayload().(*UpdateNode).WriteTo(w)
+		if err != nil {
+			return false
+		}
+		n += n2
+		return true
```

Comment:
Is this a closure?

Reply:
No. When return true, it goes next. When return false, it stops.

Reply:
Yes, it's a closure.